src/python/pants/backend/codegen/protobuf/go/rules.py

```diff
@@ -54,11 +80,286 @@ class _SetupGoProtocPlugin:
+        raise ValueError(
+            f"Failed to analyze Go sources generated from {request.import_path}.\n\n"
+            "This may be a bug in Pants. Please report this issue at "
+            "https://github.com/pantsbuild/pants/issues/new/choose and include the following data: "
+            f"error:\n{error}"
+        )
+    analysis = FirstPartyPkgAnalysis(
+        dir_path=gen_dir,
+        import_path=request.import_path,
+        imports=tuple(metadata.get("Imports", [])),
+        test_imports=tuple(metadata.get("TestImports", [])),
+        xtest_imports=tuple(metadata.get("XTestImports", [])),
+        go_files=tuple(metadata.get("GoFiles", [])),
+        test_files=tuple(metadata.get("TestGoFiles", [])),
+        xtest_files=tuple(metadata.get("XTestGoFiles", [])),
+        s_files=tuple(metadata.get("SFiles", [])),
+        minimum_go_version="",  # TODO: Get this from go.mod or elsewhere?
+        embed_patterns=tuple(metadata.get("EmbedPatterns", [])),
+        test_embed_patterns=tuple(metadata.get("TestEmbedPatterns", [])),
+        xtest_embed_patterns=tuple(metadata.get("XTestEmbedPatterns", [])),
+    )
+
+    # Obtain build requests for third-party dependencies.
+    # TODO: Consider how to merge this code with existing dependency inference code.
+    dep_build_request_addrs: list[Address] = []
+    for dep_import_path in (*analysis.imports, *analysis.test_imports, *analysis.xtest_imports):
+        # Infer dependencies on other Go packages.
+        candidate_addresses = package_mapping.mapping.get(dep_import_path)
+        if candidate_addresses:
+            # TODO: Use explicit dependencies to disambiguate? This should never happen with Go backend though.
+            if len(candidate_addresses) > 1:
+                raise ValueError(
+                    f"Multiple addresses match import of `{dep_import_path}`.\n"
+                    f"addresses: {', '.join(str(a) for a in candidate_addresses)}"
+                )
+            dep_build_request_addrs.extend(candidate_addresses)
+
+        # Infer dependencies on other generated Go sources.
+        go_protobuf_candidate_addresses = go_protobuf_mapping.mapping.get(dep_import_path)
+        if go_protobuf_candidate_addresses:
+            dep_build_request_addrs.extend(go_protobuf_candidate_addresses)
+
+    dep_build_requests = await MultiGet(
+        Get(BuildGoPackageRequest, BuildGoPackageTargetRequest(addr))
+        for addr in dep_build_request_addrs
+    )
+
+    return BuildGoPackageRequest(
+        import_path=request.import_path,
+        digest=gen_sources.digest,
+        dir_path=analysis.dir_path,
+        go_file_names=analysis.go_files,
+        s_file_names=analysis.s_files,
+        direct_dependencies=dep_build_requests,
+        minimum_go_version=analysis.minimum_go_version,
+    )
+
+
 @rule
 async def setup_build_go_package_request_for_protobuf(
-    _: GoCodegenBuildProtobufRequest,
+    request: GoCodegenBuildProtobufRequest,
```

Comment:
@Eric-Arellano: Relative to #14751, I'm still not quite sure why Go needs a union separate from the `GenerateSourcesRequest` union, which already indicates that code can be generated for Go.

A lot of work went in here, and I won't block. But it would be good to be sure that it is necessary relative to using an existing `BuildGoPackageRequest` with a target which "can be converted to" Go, rather than one that already has matching `FieldSets` (i.e.: [this code becoming codegen aware](https://github.com/pantsbuild/pants/blob/b73328f4af9eea24a63b55578f6c4e5a67c3d530/src/python/pants/backend/go/util_rules/build_pkg_target.py#L124) rather than using direct field access).

The inference `@rules` (on the other hand) do seem necessarily custom, since they map from protobuf sources to a Go path.